File: utils/cpp/test_data_loader.cpp
/**
 * @file test_data_loader.cpp
 * @brief Test Data Loader - Implementation for loading test_data.json files.
 * 
 * This module provides a unified interface for loading test data across all
 * C++ sender implementations, eliminating code duplication and providing consistent
 * path resolution.
 */
#include "test_data_loader.hpp"
#include "json.hpp"
#include <sstream>
#include <stdexcept>

using json = nlohmann::json;

namespace test_data_loader {
    
    std::filesystem::path getDefaultTestDataPath() {
        // List of potential locations to search
        std::filesystem::path search_paths[] = {
            std::filesystem::current_path(),                    // Current working directory
            std::filesystem::current_path().parent_path(),      // Parent directory (repo root)
            std::filesystem::path("/home/tim/repos")            // Absolute path
        };
        
        for (const auto& base_path : search_paths) {
            std::filesystem::path test_data_path = base_path / "test_data.json";
            if (std::filesystem::exists(test_data_path)) {
                return test_data_path;
            }
        }
        
        // If not found in search paths, return default location in /home/tim/repos
        return std::filesystem::path("/home/tim/repos") / "test_data.json";
    }
    
    std::filesystem::path resolveTestDataPath(const std::string& data_path) {
        std::filesystem::path path;
        
        if (!data_path.empty()) {
            // Use provided path
            path = std::filesystem::path(data_path);
            if (!path.is_absolute()) {
                // If relative, make it relative to current working directory
                path = std::filesystem::current_path() / path;
            }
        } else {
            // Use default path resolution
            path = getDefaultTestDataPath();
        }
        
        // Verify the file exists
        if (!std::filesystem::exists(path)) {
            std::ostringstream oss;
            oss << "test_data.json not found at: " << path.string();
            throw std::runtime_error(oss.str());
        }
        
        // Check if it's a file
        if (!std::filesystem::is_regular_file(path)) {
            std::ostringstream oss;
            oss << "Path is not a file: " << path.string();
            throw std::runtime_error(oss.str());
        }
        
        return path;
    }
    
    std::vector<json> loadTestData(const std::string& data_path) {
        try {
            std::filesystem::path resolved_path = resolveTestDataPath(data_path);
            
            std::ifstream file(resolved_path);
            if (!file.is_open()) {
                std::ostringstream oss;
                oss << "Failed to open test data file: " << resolved_path.string();
                throw std::runtime_error(oss.str());
            }
            
            json test_data;
            file >> test_data;
            file.close();
            
            return test_data.get<std::vector<json>>();
            
        } catch (const json::parse_error& e) {
            std::ostringstream oss;
            oss << "Invalid JSON in test data file: " << e.what();
            throw std::runtime_error(oss.str());
        } catch (const std::runtime_error&) {
            // Re-throw runtime errors from resolveTestDataPath
            throw;
        } catch (const std::exception& e) {
            std::ostringstream oss;
            oss << "Failed to load test data: " << e.what();
            throw std::runtime_error(oss.str());
        }
    }
    
    size_t getTestDataCount(const std::string& data_path) {
        std::filesystem::path resolved_path = resolveTestDataPath(data_path);
        
        std::ifstream file(resolved_path);
        if (!file.is_open()) {
            std::ostringstream oss;
            oss << "Failed to open test data file: " << resolved_path.string();
            throw std::runtime_error(oss.str());
        }
        
        json test_data;
        file >> test_data;
        file.close();
        
        return test_data.size();
    }
    
    std::pair<bool, std::vector<std::string>> validateTestData(const std::vector<json>& test_data) {
        std::vector<std::string> issues;
        
        for (size_t i = 0; i < test_data.size(); ++i) {
            const auto& msg = test_data[i];
            
            if (!msg.is_object()) {
                issues.push_back("Message " + std::to_string(i) + " is not a dictionary");
                continue;
            }
            
            // Check for common required fields
            if (!msg.contains("message_id")) {
                issues.push_back("Message " + std::to_string(i) + " is missing 'message_id' field");
            }
            
            if (!msg.contains("target")) {
                issues.push_back("Message " + std::to_string(i) + " is missing 'target' field");
            }
            
            if (!msg.contains("payload")) {
                issues.push_back("Message " + std::to_string(i) + " is missing 'payload' field");
            }
        }
        
        return {issues.empty(), issues};
    }
    
    void printUsage() {
        std::cout << "Test Data Loader - Simple Test" << std::endl;
        std::cout << "==============================" << std::endl;
        std::cout << std::endl;
        std::cout << "Usage:" << std::endl;
        std::cout << "  1. Include test_data_loader.hpp in your code" << std::endl;
        std::cout << "  2. Call loadTestData() to load test data" << std::endl;
        std::cout << std::endl;
        std::cout << "Example:" << std::endl;
        std::cout << "  #include \"test_data_loader.hpp\"" << std::endl;
        std::cout << "  auto test_data = test_data_loader::loadTestData();" << std::endl;
        std::cout << std::endl;
        std::cout << "Or with custom path:" << std::endl;
        std::cout << "  auto test_data = test_data_loader::loadTestData(\"/path/to/data.json\");" << std::endl;
    }
    
} // namespace test_data_loader

// Simple standalone test when compiled directly
#ifdef TEST_DATA_LOADER_STANDALONE

int main() {
    test_data_loader::printUsage();
    std::cout << std::endl;
    
    try {
        // Try to load test data
        auto test_data = test_data_loader::loadTestData();
        std::cout << "Successfully loaded " << test_data.size() << " messages" << std::endl;
        
        // Validate structure
        auto [is_valid, issues] = test_data_loader::validateTestData(test_data);
        if (is_valid) {
            std::cout << "Test data structure is valid" << std::endl;
        } else {
            std::cout << "Validation issues found:" << std::endl;
            for (const auto& issue : issues) {
                std::cout << "  - " << issue << std::endl;
            }
        }
        
        // Show sample message
        if (!test_data.empty()) {
            std::cout << "\nSample message (first item):" << std::endl;
            std::cout << test_data[0].dump(2) << std::endl;
        }
        
    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << std::endl;
        return 1;
    }
    
    return 0;
}

#endif // TEST_DATA_LOADER_STANDALONE

File: utils/python/test_data_loader.py
#!/usr/bin/env python3
"""
Test Data Loader - Centralized library for loading test_data.json files.

This module provides a unified interface for loading test data across all
sender implementations, eliminating code duplication and providing consistent
path resolution.
"""
import json
import os
from pathlib import Path
from typing import Dict, List, Optional, Any


# Default test data file name
DEFAULT_TEST_DATA_FILE = "test_data.json"


def get_default_test_data_path() -> Path:
    """
    Get the default path to test_data.json by searching common locations.
    
    Searches in the following order:
    1. Current working directory
    2. Parent directory (repo root)
    3. /home/tim/repos directory (absolute path)
    
    Returns:
        Path: The path to test_data.json
    """
    # List of potential locations to search
    search_paths = [
        Path.cwd(),                    # Current working directory
        Path.cwd().parent,             # Parent directory (repo root)
        Path("/home/tim/repos"),       # Absolute path
    ]
    
    for base_path in search_paths:
        test_data_path = base_path / DEFAULT_TEST_DATA_FILE
        if test_data_path.exists():
            return test_data_path
    
    # If not found in search paths, return default location in /home/tim/repos
    return Path("/home/tim/repos") / DEFAULT_TEST_DATA_FILE


def resolve_test_data_path(data_path: Optional[str] = None) -> Path:
    """
    Resolve the test data file path.
    
    Args:
        data_path: Optional custom path. If provided, uses this path.
                   If None, searches for default test_data.json.
    
    Returns:
        Path: The resolved path to test_data.json
    
    Raises:
        FileNotFoundError: If the file cannot be found at the specified or default location.
        PermissionError: If the file exists but cannot be read.
    """
    if data_path:
        # Use provided path
        path = Path(data_path)
        if not path.is_absolute():
            # If relative, make it relative to current working directory
            path = Path.cwd() / path
    else:
        # Use default path resolution
        path = get_default_test_data_path()
    
    # Verify the file exists and is readable
    if not path.exists():
        raise FileNotFoundError(f"test_data.json not found at: {path}")
    
    if not path.is_file():
        raise ValueError(f"Path is not a file: {path}")
    
    # Check read permissions
    if not os.access(path, os.R_OK):
        raise PermissionError(f"Cannot read file: {path}")
    
    return path


def load_test_data(data_path: Optional[str] = None) -> List[Dict[str, Any]]:
    """
    Load test data from a JSON file.
    
    This is the main function used by senders to load test data.
    It handles path resolution, file opening, and JSON parsing.
    
    Args:
        data_path: Optional path to the test data file. If not provided,
                   searches for test_data.json in common locations.
    
    Returns:
        List[Dict]: A list of message dictionaries loaded from the JSON file.
    
    Raises:
        FileNotFoundError: If the test data file cannot be found.
        PermissionError: If the file exists but cannot be read.
        json.JSONDecodeError: If the file contains invalid JSON.
        Exception: For any other errors during loading.
    
    Example:
        >>> from test_data_loader import load_test_data
        >>> test_data = load_test_data()  # Uses default location
        >>> # or
        >>> test_data = load_test_data("/path/to/test_data.json")
        >>> print(f"Loaded {len(test_data)} messages")
    """
    try:
        resolved_path = resolve_test_data_path(data_path)
        
        with open(resolved_path, 'r') as f:
            test_data = json.load(f)
        
        return test_data
    
    except json.JSONDecodeError as e:
        raise json.JSONDecodeError(
            f"Invalid JSON in test data file: {e.msg}",
            doc=e.doc,
            pos=e.pos
        )
    except Exception as e:
        raise Exception(f"Failed to load test data: {str(e)}")


def get_test_data_count(data_path: Optional[str] = None) -> int:
    """
    Get the number of messages in the test data file without loading all data.
    
    This is a lightweight function that just counts items in the JSON array
    without fully parsing the data structure.
    
    Args:
        data_path: Optional path to the test data file.
    
    Returns:
        int: Number of messages in the test data.
    
    Raises:
        FileNotFoundError: If the test data file cannot be found.
        PermissionError: If the file exists but cannot be read.
        json.JSONDecodeError: If the file contains invalid JSON.
    """
    resolved_path = resolve_test_data_path(data_path)
    
    with open(resolved_path, 'r') as f:
        # Read just enough to count the top-level array
        content = f.read()
        data = json.loads(content)
    
    return len(data)


def validate_test_data(test_data: List[Dict[str, Any]]) -> tuple[bool, List[str]]:
    """
    Validate test data structure.
    
    Checks that the test data has the expected structure with required fields.
    
    Args:
        test_data: The test data to validate.
    
    Returns:
        tuple: (is_valid, list_of_issues)
    """
    issues = []
    
    if not isinstance(test_data, list):
        issues.append("Test data must be a list of messages")
        return False, issues
    
    for i, msg in enumerate(test_data):
        if not isinstance(msg, dict):
            issues.append(f"Message {i} is not a dictionary")
            continue
        
        # Check for common required fields
        if 'message_id' not in msg:
            issues.append(f"Message {i} is missing 'message_id' field")
        
        if 'target' not in msg:
            issues.append(f"Message {i} is missing 'target' field")
        
        if 'payload' not in msg:
            issues.append(f"Message {i} is missing 'payload' field")
    
    return len(issues) == 0, issues


# Module initialization
if __name__ == "__main__":
    # Example usage and simple test
    import sys
    
    print("Test Data Loader - Simple Test")
    print("=" * 40)
    
    try:
        # Try to load test data
        test_data = load_test_data()
        print(f"✓ Successfully loaded {len(test_data)} messages")
        
        # Validate structure
        is_valid, issues = validate_test_data(test_data)
        if is_valid:
            print("✓ Test data structure is valid")
        else:
            print(f"✗ Validation issues found:")
            for issue in issues:
                print(f"  - {issue}")
        
        # Show sample message
        if test_data:
            print(f"\nSample message (first item):")
            print(json.dumps(test_data[0], indent=2))
    
    except FileNotFoundError as e:
        print(f"✗ File not found: {e}")
        sys.exit(1)
    except PermissionError as e:
        print(f"✗ Permission denied: {e}")
        sys.exit(1)
    except json.JSONDecodeError as e:
        print(f"✗ Invalid JSON: {e}")
        sys.exit(1)
    except Exception as e:
        print(f"✗ Error: {e}")
        sys.exit(1)

